### PR TITLE
Changed crawler_template.py to scrape twitter's infinite scroll.

### PR DESCRIPTION
**Pull request by dbeis**

### CODE REVIEW

File: crawler/twitter/crawler_template.py [hunk 1]
```diff
@@ -3,12 +3,51 @@
...
     print('Retrieving latest data for progress')
+    # Init config
+    interval_secs = 5
+    url = 'https://twitter.com/i/search/timeline?l=&f=tweets&q=bitcoin&src=typed&max_position='
+    min_position = ''
+    has_more_items = True
+    min_position = ''
+    
     # step one, fetch progress
+    '''
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: don't comment out, just remove

File: crawler/twitter/crawler_template.py [hunk 2]
```diff
@@ -29,22 +68,37 @@ def start_tweet_crawler(config):
...
+        
+        # Get request (XHR)
         # include error handling for twitter etc
+        try:
+            response = session.get(url + min_position, headers=headers)    
+        except requests.exceptions.Timeout:
+            raise
+            # Maybe set up for a retry, or continue in a retry loop
+        except requests.exceptions.TooManyRedirects:
+            raise
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: add discord notifications for exceptions if you cant recover, dont just raise

File: crawler/twitter/crawler_template.py [hunk 3]
```diff
@@ -63,11 +117,12 @@ def start_tweet_crawler(config):
...
         # set up for next iteration
-        latest_cid = new_data.data[-1].cid
+        min_position = json_response['min_position']
 
         # if reached end of tweets, wait some time till you get new ones (like an hour)
         # ..snip..
 
         # account for twitter's api usage metering
         # ..snip.. // sleep()
+        time.sleep(has_more_items)
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: why do you need this?

File: crawler/twitter/crawler_template.py [hunk 4]
```diff
@@ -3,12 +3,51 @@
...
 
 def start_tweet_crawler(config):
     print('\n\n')
 
     print('Retrieving latest data for progress')
+    # Init config
+    interval_secs = 5
+    url = 'https://twitter.com/i/search/timeline?l=&f=tweets&q=bitcoin&src=typed&max_position='
+    min_position = ''
+    has_more_items = True
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: this never gets updated, infninite loop

File: crawler/twitter/crawler_template.py [hunk 5]
```diff
@@ -3,12 +3,51 @@
...
+    headers['User-Agent'] = 'TweetScraper' # Mandatory
+    headers['Accept-Encoding'] = 'gzip,deflate' # Optional
 
 def start_tweet_crawler(config):
     print('\n\n')
 
     print('Retrieving latest data for progress')
+    # Init config
+    interval_secs = 5
+    url = 'https://twitter.com/i/search/timeline?l=&f=tweets&q=bitcoin&src=typed&max_position='
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: use parameter passing

File: crawler/twitter/crawler_template.py [hunk 6]
```diff
@@ -3,12 +3,51 @@
...
+                tweet_timestamp = html_tweet.select_one('li .time span[data-time]').get('data-time'),
+                ['bitcoin']
+            )
+        )
+    
+    return crawledData_tweets
+
+# Set headers as needed
+def setHeaders(headers, url=None):
+    # headers param includes those returned from the previous request
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: not needed comments

File: crawler/twitter/crawler_template.py [hunk 7]
```diff
@@ -29,22 +68,37 @@ def start_tweet_crawler(config):
...
     # step two, data fetch-post loop
-    while(True):
+    while(has_more_items):
         print(f'Fetching more twitter data starting from cid {latest_cid}')
         # data fetch ..snip.. (omitted for brevity)
+        # Get headers and add them to next request
+        headers = session.cookies.get_dict()
+        setHeaders(headers, url + min_position)
+        
+        # Get request (XHR)
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: please remove all unneeded comments

File: crawler/twitter/crawler_template.py [hunk 8]
```diff
@@ -29,22 +68,37 @@ def start_tweet_crawler(config):
...
         # include error handling for twitter etc
+        try:
+            response = session.get(url + min_position, headers=headers)    
+        except requests.exceptions.Timeout:
+            raise
```
Comment (Zarkopafilis) [inline crawler/twitter/crawler_template.py]: discord

File: crawler/twitter/crawler.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        headers = session.cookies.get_dict()
+        set_headers(headers, url + min_position)
+        
+        # Get request (XHR)
+        # include error handling for twitter etc
+        try:
+            response = session.get(url + min_position, headers=headers)    
+        except requests.exceptions.Timeout as e:
+            print(str(e))
+            # if timeout error happens, notify discord
```
Comment (deepettas) [inline crawler/twitter/crawler.py]: Remove redundant comments like this

File: crawler/twitter/crawler.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+            notify('crawler', warning_update(str(e), 'twitter'))
+            continue
+            # Maybe set up for a retry, or continue in a retry loop
+        except requests.exceptions.TooManyRedirects as e:
+            print(str(e))
+            notify('crawler', error_update(str(e), 'twitter'))
+            return
+            # Tell the user their URL was bad and try a different one
```
Comment (deepettas) [inline crawler/twitter/crawler.py]: Redundant comment

**Comment by Zarkopafilis** [comment]
I'm merging this after discussion with @deepettas . @dbeis please use github 'resolve changes' for tracking this better next time. Will open tasks to unit test this, further abstract tags (bitcoin, twitter urls etc).

Thanks for contributing.

**Pull request merged**